src/user_database.ts

```diff
@@ -365,13 +388,24 @@ export class KnexUserDatabase implements UserDatabase {
     return result;
   }
 
-  async query<R>(sql: string, ...params: any[]): Promise<R[]> {
+  async queryFunction<C extends UserDatabaseClient, R, T extends unknown[]>(func: UserDatabaseQuery<C, R, T>, ...args: T): Promise<R> {
+    const result = await this.knex.transaction<R>(
+      async (transactionClient: Knex.Transaction) => {
+        return await func(transactionClient as unknown as C, ...args);
+      },
+      { isolationLevel: "read committed", readOnly : true }
```

Comment:
Do we want to use read-only transaction this for pg-node, TypeORm, and Prisma client as well?

Reply:
I'm not sure.  I would establish the API and be willing to change the implementation later.
FWIW, I did not add pg-node... that's deprecated right?

Reply:
It seems to me, also, that some robust applications count failed attempts and react... that might count as logging but it might also be kept in in-memory or durable state.

Reply:
After more consideration, I think at some point we will be expected to introduce database users/roles in some way.  The user/role for deployment, regular operation, and the auth middleware may all be different.  (But the API will be the same.)
I don't expect to do this in this PR, but I think it is a better way forward than worrying about read-only.

Reply:
I think `const client: PoolClient = await this.pool.connect();` is the pg-node support, but we don't expect end users to directly use pg-node.